### ISSUE TITLE
tcg_prologue_init() overrides code_gen_buffer

### ISSUE THREAD

**Issue by PhilippTakacs** [unicorn-engine/unicorn#1567]
Hi

The following code overrides code_gen_buffer of the TCGContext:
https://github.com/unicorn-engine/unicorn/blob/c10639fd4658a852049546162d116b123e2b1ec2/qemu/tcg/tcg.c#L849-L857

Similar code is also in tcg_region_assign().

code_gen_buffer is assigned in translate-all.c:
https://github.com/unicorn-engine/unicorn/blob/c10639fd4658a852049546162d116b123e2b1ec2/qemu/accel/tcg/translate-all.c#L955

alloc_code_gen_buffer() gets the buffer depending on the platform and build time options from static memory, VirtualAlloc() or mmap(). Later free_code_gen_buffer() is called to de allocate the buffer. On linux the munmap will fail with EINVAL and the page is still mapped. In the end this is a memory leak, because after a uc_close() nobody has access to the page.

**Comment by wtdcode** [comment]
Seems a good catch. I will write an ASAN for it.

**Comment by PhilippTakacs** [comment]
> I will write an ASAN for it

Not sure what you mean with this. I have tried to find this bug with -fsanitize=address but turns out the sanitizer can't detect mmap leaks (https://github.com/google/sanitizers/issues/1224). In the end I found this with strace and a debugger. To find this with a sanitizer a alloc_code_gen_buffer() implementation with malloc() could work.

To the bug itself: Is there any documentation for the TCGContext? If I would understand how the struct is supposed to work I could write a patch.
